Add docs for gossip encryption key rotation (#10522)
hashicorp/consul: website/content/docs/k8s/operations/gossip-encryption-key-rotation.mdx

---
layout: docs
page_title: Gossip Encryption Key Rotation
description: Rotate the Gossip Encryption Key on Kubernetes Cluster safely
---

# Rotating Gossip Encryption Key

> Follow these steps update the gossip encryption key on the Consul cluster.

Reviewer: ```suggestion
The following instructions provides a step by step manual process for rotating gossip encryption keys on Consul Servers that are deployed onto a Kubernetes cluster with Consul on Kubernetes.
```

> The following instructions provides a step-by-step manual process for rotating gossip encryption keys on Consul Servers that are deployed onto a Kubernetes cluster with Consul on Kubernetes.

Reviewer 2: Should link to something that describes what gossip encryption keys are

Reviewer 2: >on Consul Servers

Don't we also need to rotate clients?

Author: sigh! i mis-wrote it. I wanted to say perform these operation on the consul server pod, but i dont think that actually matters. Let me correct that.

  **Note:** In case of federated Consul clusters, perform the following steps in the primary datacenter.

1. Generate a new key:

  ```shell-session
  consul keygen
  ```
  This should generate a new key which can be used as the gossip encryption key. In this example, we will be using
  `Wa6/XFAnYy0f9iqVH2iiG+yore3CqHSemUy4AIVTa/w=` as the replacement gossip encryption key.

2. Add new key to consul keyring.

>   `kubectl exec` into a Consul Server pod and add the new key to the Consul Keyring. This can be performed by running the following command:

Reviewer 2: If any server pod works (does't have to be the leader) then we should say that explicitly

  ```shell-session

>   consul keyring -install="Wa6/XFAnYy0f9iqVH2iiG+yore3CqHSemUy4AIVTa/w="

Reviewer 2: Does this require ACLs? If so we should describe which ACL token to use.

  ==> Installing new gossip encryption key...
  ```

>   List the keys in the keyring to verify the new key has been installed successfully.

Reviewer 2: According to https://learn.hashicorp.com/tutorials/consul/gossip-encryption-rotate#verify-that-the-key-was-installed it's important to ensure that all of these numbers are `n/n`. I think that needs to be explicitly called out.

Reviewer 2: We should also state here that Consul is automatically propagating this key to all clients and servers.

  ```shell-session
  consul keyring -list
  ==> Gathering installed encryption keys...

  WAN:
    CL6M+jKj3630CZLXI0IRVeyci1jgIAveiZKvdtTybbA= [2/2]
    Wa6/XFAnYy0f9iqVH2iiG+yore3CqHSemUy4AIVTa/w= [2/2]

  dc1 (LAN):
    CL6M+jKj3630CZLXI0IRVeyci1jgIAveiZKvdtTybbA= [4/4]
    Wa6/XFAnYy0f9iqVH2iiG+yore3CqHSemUy4AIVTa/w= [4/4]

  dc2 (LAN):
    CL6M+jKj3630CZLXI0IRVeyci1jgIAveiZKvdtTybbA= [4/4]
    Wa6/XFAnYy0f9iqVH2iiG+yore3CqHSemUy4AIVTa/w= [4/4]
  ```

3. Use the new key as the gossip encryption key.

  Once the new key has been added to the keychain, you can install it as the new gossip encryption key. In order to do so

>   run the following while remaining `kubectl exec`-d in the Consul server pod.

Reviewer 3: I wonder if this looks cuter? 
```suggestion
  run the following while remaining `kubectl exec`'d in the Consul server pod.
```

Meta comment: I'm curious if you need to set any CONSUL_HTTP_TOKEN/other exports, doesn't seem like it, I wonder if the `consul keyring` command *should* require some escalated privileges?

Author: Yeah. we do need to export the `CONSUL_HTTP_TOKEN` if ACLs are enabled. Ill mention that.

Reviewer 3: OK cool, I thought so, it's not a blocker by any means but maybe worth a call-out since someone will be scripting based on this :)

  ```shell-session
  consul keyring -use="Wa6/XFAnYy0f9iqVH2iiG+yore3CqHSemUy4AIVTa/w="
  ==> Changing primary gossip encryption key...
  ```

> 4. Update the kubernetes secrets with the latest encryption key.

Reviewer: ```suggestion
4. Update the Kubernetes secrets with the latest gossip encryption key.
```

>   In order for subsequent `helm upgrades` to occur without unexpected behavior, it is important to update the secrets with

Reviewer: ```suggestion
  In order for subsequent `helm upgrade` commands to be successfully executed without unexpected behavior, it is important to update the secrets with
```

>   In order for subsequent `helm upgrades` commands to be successfully executed without unexpected behavior, it is important to update the secrets with

Reviewer 2: ```suggestion
  In order for subsequent `helm upgrades` commands to be successfully executed without unexpected behavior, it is important to update the gossip encryption Kubernetes Secret with
```

  the value of the new gossip encryption key.

>   The name of the secret which stores the value of the gossip encryption key can be found in the helm values file.

Reviewer: ```suggestion
  The name of the secret which stores the value of the gossip encryption key can be found in the Helm values file.
```

>   The name of the secret which stores the value of the gossip encryption key can be found in the Helm values file.

Reviewer 2: ```suggestion
  The name of the secret which stores the value of the gossip encryption key can be found in the Helm values file:
```

  ```yaml
  global:
    gossipEncryption:
       secretName: consul-gossip-encryption-key
       secretKey: key
  ```

>   As the values of Kubernetes secrets are `base64` encoded, encode the gossip encryption key to base 64.

Reviewer 2: We can use `stringData` to avoid the base64 encoding which will help because the `base64` may have different options on different platforms.

```
 kubectl patch secret consul-gossip-key -p='{"stringData":{"key": "Wa6/XFAnYy0f9iqVH2iiG+yore3CqHSemUy4AIVTa/w="}}'
```

  ```shell-session
  echo "Wa6/XFAnYy0f9iqVH2iiG+yore3CqHSemUy4AIVTa/w=" | base64 -i -
  V2E2L1hGQW5ZeTBmOWlxVkgyaWlHK3lvcmUzQ3FIU2VtVXk0QUlWVGEvdz0K

  kubectl patch secret consul-gossip-encryption-key -p '{"data":{"key": "V2E2L1hGQW5ZeTBmOWlxVkgyaWlHK3lvcmUzQ3FIU2VtVXk0QUlWVGEvdz0K"}}'
  ```

>   **Note:** In case of federated Consul clusters, update the federation-secret value for the gossip encryption key. This name of the secret and key can be found in the values file of the secondary datacenter.

Reviewer: ```suggestion
  **Note:** In the case of federated Consul clusters, update the federation-secret value for the gossip encryption key. This name of the secret and key can be found in the values file of the secondary datacenter.
```

>   **Note:** In the case of federated Consul clusters, update the federation-secret value for the gossip encryption key. This name of the secret and key can be found in the values file of the secondary datacenter.

Reviewer 3: ```suggestion
  **Note:** In the case of federated Consul clusters, update the federation-secret value for the gossip encryption key. The name of the secret and key can be found in the values file of the secondary datacenter.
```

  ```yaml
  global:
     gossipEncryption:
       secretName: consul-federation
       secretKey: gossipEncryptionKey
  ```

  ```shell-session
  kubectl patch secret consul-federation -p '{"data":{"gossipEncryptionKey": "V2E2L1hGQW5ZeTBmOWlxVkgyaWlHK3lvcmUzQ3FIU2VtVXk0QUlWVGEvdz0K"}}'

>   ```

Reviewer 2: Should we add a step to remove the old key? (https://learn.hashicorp.com/tutorials/consul/gossip-encryption-rotate#remove-the-old-key-from-the-keyring). That's probably why folks are rotating it, so the old secret is no longer usable.
